src/librbd/image/CreateRequest.cc

```diff
@@ -0,0 +1,678 @@
+void CreateRequest<I>::fetch_mirror_image() {
+  if ((m_mirror_mode != RBD_MIRROR_MODE_POOL) && !m_force_non_primary) {
+    complete(0);
+    return;
+  }
+
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
+
+  librados::ObjectReadOperation op;
+  cls_client::mirror_image_get_start(&op, m_image_ctx->id);
+
+  using klass = CreateRequest<I>;
+  librados::AioCompletion *comp =
+    create_rados_ack_callback<klass, &klass::handle_fetch_mirror_image>(this);
+  m_outbl.clear();
+  int r = m_image_ctx->md_ctx.aio_operate(RBD_MIRRORING, comp, &op, &m_outbl);
+  assert(r == 0);
+  comp->release();
+}
+
+template<typename I>
+Context *CreateRequest<I>::handle_fetch_mirror_image(int *result) {
+  ldout(m_cct, 20) << __func__ << ": r=" << *result << dendl;
+
+  // this should always be the case during create
+  assert(*result == -ENOENT);
+
+  mirror_image_enable();
+  return nullptr;
+}
+
+template<typename I>
+void CreateRequest<I>::mirror_image_enable() {
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
+
+  m_mirror_image_internal.state = cls::rbd::MIRROR_IMAGE_STATE_ENABLED;
+  if (m_non_primary_global_image_id.empty()) {
+    uuid_d uuid_gen;
+    uuid_gen.generate_random();
+    m_mirror_image_internal.global_image_id = uuid_gen.to_string();
+  } else {
+    m_mirror_image_internal.global_image_id = m_non_primary_global_image_id;
+  }
+
+  librados::ObjectWriteOperation op;
+  cls_client::mirror_image_set(&op, m_image_ctx->id, m_mirror_image_internal);
+
+  using klass = CreateRequest<I>;
+  librados::AioCompletion *comp =
+    create_rados_ack_callback<klass, &klass::handle_mirror_image_enable>(this);
+  int r = m_image_ctx->md_ctx.aio_operate(RBD_MIRRORING, comp, &op);
+  assert(r == 0);
+  comp->release();
+}
+
+template<typename I>
+Context *CreateRequest<I>::handle_mirror_image_enable(int *result) {
+  ldout(m_cct, 20) << __func__ << ": r=" << *result << dendl;
+
+  if (*result < 0) {
+    lderr(m_cct) << "cannot enable mirroring: " << cpp_strerror(*result) << dendl;
+
+    m_r_saved = *result;
+    CreateRequest<I>::journal_remove();
+    return nullptr;
+  }
+
+  send_watcher_notification();
+  return nullptr;
+}
+
+template<typename I>
+void CreateRequest<I>::send_watcher_notification() {
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
+
+  Context *ctx = new FunctionContext([this](int r) {
+      r = MirroringWatcher<>::notify_image_updated(
+        m_image_ctx->md_ctx, cls::rbd::MIRROR_IMAGE_STATE_ENABLED,
+        m_image_ctx->id, m_mirror_image_internal.global_image_id);
+      CreateRequest<I>::handle_watcher_notify(r);
+    });
+
+    m_image_ctx->op_work_queue->queue(ctx, 0);
+}
+
+template<typename I>
+void CreateRequest<I>::handle_watcher_notify(int r) {
+  ldout(m_cct, 20) << __func__ << ": r=" << r << dendl;
+
+  if (r < 0) {
+    lderr(m_cct) << "failed to send update notification: " << cpp_strerror(r)
+                 << dendl;
+
+    m_r_saved = r;
+    CreateRequest<I>::mirror_image_remove();
```

Comment:
I don't think it's worth failing the create if one-or-more watchers failed to ack it in time.  They will pick up the update when they notice they missed an update (and rescan the full mirroring directory).